File: tests/domains/video/test_thumbnail.py
import pytest

from domains import Runtime
from domains.video.probe import probe_video
from domains.video.thumbnail import create_thumbnail
from tests.helpers import unwrap

pytestmark = pytest.mark.ffmpeg


def test_create_thumbnail_youtube(runtime: Runtime, sample_video: str) -> None:
    result = unwrap(create_thumbnail(runtime, sample_video, title="Título de teste grande"))
    assert result["output"] == "sample_thumb_youtube.jpg"
    assert (result["width"], result["height"]) == (1280, 720)
    assert result["time"] == 1.5
    assert result["size_bytes"] > 0
    probed = unwrap(probe_video(runtime, result["output"]))
    assert probed["info"]["width"] == 1280


def test_create_thumbnail_vertical_custom(runtime: Runtime, sample_video: str) -> None:
    result = unwrap(
        create_thumbnail(
            runtime,
            sample_video,
            time=0.2,
            title="Sem darken",
            thumbnail_format="vertical",
            position="bottom",
            text_color="#FFD700",
            darken=False,
            output="capas/capa.png",
        )
    )
    assert result["output"] == "capas/capa.png"
    assert (result["width"], result["height"]) == (1080, 1920)


def test_create_thumbnail_from_image(runtime: Runtime, sample_video: str) -> None:
    base = unwrap(create_thumbnail(runtime, sample_video, thumbnail_format="original"))
    assert (base["width"], base["height"]) == (320, 240)
    result = unwrap(create_thumbnail(runtime, base["output"], title="Capa"))
    assert result["time"] is None


def test_create_thumbnail_invalid(runtime: Runtime, sample_video: str) -> None:
    assert create_thumbnail(runtime, sample_video, time=99).get("code") == "invalid_argument"
    assert create_thumbnail(runtime, sample_video, title="x" * 81).get("code") == (
        "invalid_argument"
    )
    assert create_thumbnail(runtime, sample_video, text_color="red").get("code") == (
        "invalid_argument"
    )
    assert create_thumbnail(runtime, sample_video, output="capa.gif").get("code") == (
        "invalid_argument"
    )


def test_create_thumbnail_missing(runtime: Runtime) -> None:
    assert create_thumbnail(runtime, "ghost.mp4").get("code") == "not_found"

File: version-in-rust/tests/domains/video/test_thumbnail.rs
use mcp_tools::core::errors::ErrorCode;
use mcp_tools::domains::video::probe::probe_video;
use mcp_tools::domains::video::thumbnail::{
    create_thumbnail, TextPosition, ThumbnailFormat, ThumbnailOptions,
};

use crate::conftest::{runtime, sample_video};
use crate::skip_without_ffmpeg;

#[test]
fn test_create_thumbnail_youtube() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let options = ThumbnailOptions {
        title: Some("Título de teste grande".to_string()),
        ..ThumbnailOptions::default()
    };
    let result = create_thumbnail(&rt.runtime, &sample, &options).unwrap();
    assert_eq!(result.output, "sample_thumb_youtube.jpg");
    assert_eq!((result.width, result.height), (1280, 720));
    assert_eq!(result.time, Some(1.5));
    assert!(result.size_bytes > 0);
    let probed = probe_video(&rt.runtime, &result.output).unwrap();
    assert_eq!(probed.info.width, Some(1280));
}

#[test]
fn test_create_thumbnail_vertical_custom() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let options = ThumbnailOptions {
        time: Some(0.2),
        title: Some("Sem darken".to_string()),
        thumbnail_format: ThumbnailFormat::Vertical,
        position: TextPosition::Bottom,
        text_color: "#FFD700".to_string(),
        darken: false,
        output: Some("capas/capa.png".to_string()),
        ..ThumbnailOptions::default()
    };
    let result = create_thumbnail(&rt.runtime, &sample, &options).unwrap();
    assert_eq!(result.output, "capas/capa.png");
    assert_eq!((result.width, result.height), (1080, 1920));
}

#[test]
fn test_create_thumbnail_from_image() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let base = create_thumbnail(
        &rt.runtime,
        &sample,
        &ThumbnailOptions {
            thumbnail_format: ThumbnailFormat::Original,
            ..ThumbnailOptions::default()
        },
    )
    .unwrap();
    assert_eq!((base.width, base.height), (320, 240));
    let result = create_thumbnail(
        &rt.runtime,
        &base.output,
        &ThumbnailOptions {
            title: Some("Capa".to_string()),
            ..ThumbnailOptions::default()
        },
    )
    .unwrap();
    assert_eq!(result.time, None);
}

#[test]
fn test_create_thumbnail_invalid() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let cases = [
        ThumbnailOptions {
            time: Some(99.0),
            ..ThumbnailOptions::default()
        },
        ThumbnailOptions {
            title: Some("x".repeat(81)),
            ..ThumbnailOptions::default()
        },
        ThumbnailOptions {
            text_color: "red".to_string(),
            ..ThumbnailOptions::default()
        },
        ThumbnailOptions {
            output: Some("capa.gif".to_string()),
            ..ThumbnailOptions::default()
        },
    ];
    for options in &cases {
        let error = create_thumbnail(&rt.runtime, &sample, options).unwrap_err();
        assert_eq!(error.code, ErrorCode::InvalidArgument);
    }
}

#[test]
fn test_create_thumbnail_missing() {
    let rt = runtime();
    let error =
        create_thumbnail(&rt.runtime, "ghost.mp4", &ThumbnailOptions::default()).unwrap_err();
    assert_eq!(error.code, ErrorCode::NotFound);
}
